### PR TITLE
Pipe operator support

### LINKED ISSUE #72: Pipe operator support
It would be great to have support for the pipe operator. It is a common way in many functional programming languages to compose functions into a readable left-to-right or top-to-bottom "pipeline" of steps of computation.

[Documentation in OCaml stdlib](https://ocaml.org/releases/4.11/htmlman/libref/Stdlib.html#VAL(|%3E))
[Documentation in Elixir Kernel](https://hexdocs.pm/elixir/master/Kernel.html#%7C%3E/2)

I think it should probably behave like in OCaml and not like Elixir, meaning this Caramel code in file `main.ml`:
```ocaml
let print_int number = Io.format "~0tp~n" [ number ]

let subtract x y = y - x
let main _ =
  let divide x y = y / x in
  10 |> subtract 2 |> divide 4 |> print_int
```
when ran with `ocaml compile main.ml && escript main.erl` would print `2` just like this Caramel code:
```ocaml
let print_int number = Io.format "~0tp~n" [ number ]

let subtract x y = y - x
let main _ =
  let divide x y = y / x in
  print_int (divide 4 (subtract 2 10))
```
as indicated by this OCaml code:
```ocaml
(* let print_int number = Io.format "~0tp~n" [ number ] *)

let subtract x y = y - x ;;
let main _ =
  let divide x y = y / x in
  10 |> subtract 2 |> divide 4 |> print_int ;;

main ()
```
when executed:
```
$ ocaml main.ml
2
```

### PR DESCRIPTION
**Pull request by ayshiff**
Closes #72

This PR adds pipe operator support (`|>`).
It is in Draft because we have to wait for #44 to be resolved first. (As mentionned [here](https://github.com/AbstractMachinesLab/caramel/issues/72#issuecomment-791216993))

### CODE REVIEW

File: stdlib/caramel_runtime.erl [hunk 1]
```diff
@@ -16,3 +17,5 @@ recv(Timeout) ->
 
 binary_concat(A, B) when is_binary(A) and is_binary(B) ->
   << (A)/binary, (B)/binary >>.
+
+pipe(A, B) -> B A.
```
Comment (michallepicki) [inline stdlib/caramel_runtime.erl:21]:
```suggestion
pipe(A, B) -> B(A).
```

I think Erlang needs the parentheses here :)

**Review by michallepicki — COMMENTED** [review]
I think this is a good step forward to get things working, great job!

File: stdlib/caramel_runtime.erl (hunk 1 above)
Comment (michallepicki) [reply stdlib/caramel_runtime.erl:21]:
```
$ erl
1> F = fun (X) -> X + 1 end.
#Fun<erl_eval.44.79398840>
2> F(1).                    
2
3> F 1.                     
* 1: syntax error before: 1
3> F 1 . 
* 1: syntax error before: 1
3> 
```

File: stdlib/caramel_runtime.ml [hunk 2]
```diff
@@ -3,3 +3,5 @@ external recv_with_timeout : int -> 'a = "recv"
 external recv_and_wait : unit -> 'a = "recv"
 
 external ( ^ ) : string -> string -> string = "binary_concat"
+
+external ( |> ) : 'a -> 'b -> 'c = "pipe"
```
Comment (ostera) [inline stdlib/caramel_runtime.ml:7]:
```suggestion
external ( |> ) : 'a -> ('a -> 'b) -> 'b = "pipe"
```

This should be a stricter signature that matches what we want out of this operator.

**Review by ostera — APPROVED** [review]
Good work! 🤩 

Looks good to me. We should be able to optimize the pipe away later on.

**Comment by ostera** [comment]
Another thought I had is that we may want to just rewrite the pipes into SSA so we minimize the amount of fun allocations too. So we'd go from this:

```ocaml
let f x = x |> add 1 |> div 2 |> mult 3
```

To this:

```erlang
f(X) ->
  Caramel@Tmp1 = add(1, X),
  Caramel@Tmp2 = div(2, Caramel@Tmp1),
  Caramel@Tmp3 = mult(3, Caramel@Tmp2).
```

This rewrite would be an optimization during translation that happens the moment we find a `|>` in there. We may need to just leave a mark and have a second compilation pass do this optimization. We'd still need the `external` to be able to type-check correctly, but we won't actually end up creating a `caramel_runtime:pipe/2`.

It also means we can take our time to work out the quirks of partial application in the presence of optional and named arguments as pointed out here: https://github.com/AbstractMachinesLab/caramel/pull/92/files#r605967813

@ayshiff would you be interested in trying this approach? 🙌🏽

**Comment by ayshiff** [comment]
> Another thought I had is that we may want to just rewrite the pipes into SSA so we minimize the amount of fun allocations too. So we'd go from this:
> 
> ```ocaml
> let f x = x |> add 1 |> div 2 |> mult 3
> ```
> 
> To this:
> 
> ```erlang
> f(X) ->
>   Caramel@Tmp1 = add(1, X),
>   Caramel@Tmp2 = div(2, Caramel@Tmp1),
>   Caramel@Tmp3 = mult(3, Caramel@Tmp2).
> ```
> 
> This rewrite would be an optimization during translation that happens the moment we find a `|>` in there. We may need to just leave a mark and have a second compilation pass do this optimization. We'd still need the `external` to be able to type-check correctly, but we won't actually end up creating a `caramel_runtime:pipe/2`.
> 
> It also means we can take our time to work out the quirks of partial application in the presence of optional and named arguments as pointed out here: https://github.com/AbstractMachinesLab/caramel/pull/92/files#r605967813
> 
> @ayshiff would you be interested in trying this approach? 🙌🏽

Yes, absolutely! I am interested in trying this 👍

**Comment by ayshiff** [comment]
Hey @ostera @michallepicki 👋
I'm here to ask for some help. I tried to follow Leandro's recommendations but I'm a bit lost.
I made a branch with my progress if you want to have a look 👍

https://github.com/ayshiff/caramel/tree/feature/pipe-operator-test

I also have a question about testing.
To generate a unique variable name in the transformation phase, I wanted to rely on random hash generation. The issue is that I don't know how we could test this behavior.

**Comment by ostera** [comment]
Hi! 👋🏽  since I'm currently not working on and have no plans to continue working on this version of the compiler, I'll close this PR. Thanks a lot for your contribution 🙏🏽

**Pull request closed without merging**